In [14]:
pip install beautifulsoup4

     |████████████████████████████████| 112kB 4.8MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [15]:
pip install lxml

     |████████████████████████████████| 5.8MB 5.5MB/s eta 0:00:01     |█                               | 194kB 1.8MB/s eta 0:00:04     |███▋                            | 645kB 1.8MB/s eta 0:00:03     |████████▉                       | 1.6MB 1.8MB/s eta 0:00:03     |██████████▋                     | 1.9MB 1.8MB/s eta 0:00:03     |█████████████                   | 2.4MB 1.8MB/s eta 0:00:02     |█████████████▉                  | 2.5MB 1.8MB/s eta 0:00:02     |█████████████████████▏          | 3.8MB 1.8MB/s eta 0:00:02
Note: you may need to restart the kernel to use updated packages.


In [14]:
pip install wikipedia api

     |████████████████████████████████| 163kB 5.0MB/s eta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/87/2a/18/4e471fd96d12114d16fe4a446d00c3b38fb9efcb744bd31f4a
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/8e/b3/13/6fb6612d6dd2d4bc2dfa783e9be7554e7c2f276912d3cb850e
Successfully built wikipedia api
Note: you may need to restart the kernel to use updated packages.


In [ ]:
# kernel restarted

In [1]:
# Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown.
import pandas as pd 
import wikipedia as wp
from bs4 import BeautifulSoup

In [2]:
# Download url from Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
html = wp.page("List of postal codes of Canada: M").html().encode("UTF-8")

In [3]:
# Determine table index
df = pd.read_html(html, header = 0)[0]

In [7]:
# Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
df = df[df.Borough != 'Not assigned']

In [8]:
df.head()

Postal code           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [9]:
# More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
df = df.groupby(['Postal code', 'Borough'])['Neighborhood'].apply(list).apply(lambda x:', '.join(x)).to_frame().reset_index()

In [10]:
# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
for index, row in df.iterrows():
    if row['Neighborhood'] == 'Not assigned':
        row['Neighborhood'] = row['Borough']

In [15]:
# Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
df = df.dropna()
empty = 'Not assigned'

In [18]:
df

Postal code      Borough  \
0           M1B  Scarborough   
1           M1C  Scarborough   
2           M1E  Scarborough   
3           M1G  Scarborough   
4           M1H  Scarborough   
..          ...          ...   
98          M9N         York   
99          M9P    Etobicoke   
100         M9R    Etobicoke   
101         M9V    Etobicoke   
102         M9W    Etobicoke   

                                          Neighborhood  
0                                      Malvern / Rouge  
1             Rouge Hill / Port Union / Highland Creek  
2                  Guildwood / Morningside / West Hill  
3                                               Woburn  
4                                            Cedarbrae  
..                                                 ...  
98                                              Weston  
99                                           Westmount  
100  Kingsview Village / St. Phillips / Martin Grov...  
101  South Steeles / Silverstone / Humbergate / Jam...  
102                                          Northwest  

[103 rows x 3 columns]

In [19]:
# In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
df.shape

(103, 3)

This dataframe has 103 distinct Postal codes from the Toronto Canada area that service several boroughs/neighborhoods with the same names. Therefore, This dataframe should be grouped by Postal code consisting of a dataframe with 103 rows.